### ISSUE TITLE
Missing documentation of built-in template variables

### ISSUE THREAD

**Issue by djzort** [RexOps/Rex#1559]
### Describe the bug

Variables from @_ of resource mysteriously inserted in to template parameters

### Expected behavior

Ideally @_ parameters arent magically inserted in to template parameters, or its something that can be controlled.

### How to reproduce it

Create a resource
call a function from it with some parameters
they will appear in the template

### Code example

```perl
package Some::Thing;
use Rex -feature => ['1.4'];
task 'doit' => sub {
my %params;
file "/tmp/test.txt", content => template("files/stuff.txt", %params);
};
package main;
Some::Thing::doit({ hello => "there" });
```

The template will have the variable `hello = there` magically inserted and there is no way to stop it.

### Additional context

It's weird that this is wrapped in Rex::Config 

Sys info is inserted Rex::Commands::File and can be disabled with the undocumented parameter: `__no_sys_info__`

### Rex version

(R)?ex 1.13.4

### Perl version

v5.36.0

### Operating system running rex

Linux

### Operating system managed by rex

Linux

### How rex was installed?

cpan client

**Comment by djzort** [comment]
I discovered this when trying out Rex::Template::TT and started getting warnings from the names of parameters passed to the resource. It had to spelunk the code extensively to find it as its undocumented and in a strange place

**Comment by ferki** [comment]
Thanks for raising this issue, @djzort! The template variable logic should be documented indeed.

> Variables from @_ of resource[^1] mysteriously inserted in to template parameters

I can confirm making task parameters available as template variables is intentional, but not mentioned in the `Rex::Commands::File::template()` docs yet.

It was added on b0416b1710611b900284ba5a2d9d74991fc4351b, which seems to bundle a few separate logical changes together for some reason :thinking: Those task parameters may come through the CLI, or via a function-style call of the task (like in the example).

> Sys info is inserted Rex::Commands::File and can be disabled with the undocumented parameter: `__no_sys_info__`

Right, that has to be documented too.

> It's weird that this is wrapped in Rex::Config

I assume this means [`get_template_function`](https://github.com/RexOps/Rex/blob/36b9d62c0d0f895f9e2d88ea087508b48c3f4156/lib/Rex/Config.pm#L1435-L1490).

Which certainly could use some refactoring to make it less repetitive, and then perhaps all the template variable compilation could be moved to a place like [`_get_std_template_vars`](https://github.com/RexOps/Rex/blob/36b9d62c0d0f895f9e2d88ea087508b48c3f4156/lib/Rex/Config.pm#L1435-L1490) or similar – where any additional logic may be handled better as well.

I expect this would leave the Rex::Config bits about configuration things only (that is "which function call to be used to render templates?"), as it should be.

> Ideally @_ parameters arent magically inserted in to template parameters, or its something that can be controlled.

Making task parameters available to templates is an intended feature, so I don't consider this part of the story as a bug.

Adding extra controls to make this behavior optional sounds like a good idea for a new feature, and I would leave it to the person working on this issue to include it in this scope or not. Perhaps it would be best to split that off into a separate issue if that fits their work better due to e.g. complexity.

Missing documentation for the template variables is certainly a bug, and has to be fixed :+1:

> I discovered this when trying out Rex::Template::TT

Using Rex::Template::TT is a good choice :+1:

[^1]: Please note that the term "resource" is used within Rex context to refer more or less to "something that can have a state ensured". What we see here in the example is a task call through its equivalent auto-registered functional form.

**Comment by djzort** [comment]
The trick with Rex::Template::TT is that it it will call its own validate_vars and it will complain about *all* variables passed in to the resource - this is the scenario that caused me to start diving. So keys with hyphens, dots or underscores will all start logging warnings.

https://metacpan.org/dist/Rex-Template-TT/source/lib/Rex/Template/TT.pm#L24

**Comment by djzort** [comment]
I created this issue by clicking on the line in the code view here on GH and following the prompts. It seems though that a link to the line in Rex::Config was not included in the issue automatically

Here is where Rex::Config wraps the template function and magically inserts all function variables. Which is in addition Rex::Command::File wrapping the template function too.

https://github.com/RexOps/Rex/blob/36b9d62c0d0f895f9e2d88ea087508b48c3f4156/lib/Rex/Config.pm#L1435

**Comment by ferki** [comment]
> [...] Rex::Template::TT [...] will complain about _all_ variables passed in [...] So keys with hyphens, dots or underscores will all start logging warnings.

Thanks for the additional context.

I can't reproduce this scenario on my systems, and I can only get a warning if I pass a non-compatible name on purpose (which is then expected to warn).

For reference, Rex::Template::TT should print a warning if a name of a passed template variable:

- starts with a dot or underscore
- contains a character other than `qr{[a-zA-Z0-9_]}`
- matches exactly with a reserved word of Template

It also sounds like a different issue than "missing docs for built-in template variables".

If you have specific examples from your scenarios where a non-compatible template variable name is coming from the _built-in_ ones, please report it as a separate bug.

If you suspect Rex::Template::TT is misbehaving, please report it in [its separate tracker](https://github.com/RexOps/Rex-Template-TT/issues).

**Comment by ferki** [comment]
I merged #1574 to fix the immediate bug about lack of documentation.

Further modifications are possible, but they seem to inflate the scope too much to my taste. I find it is only possible to fully specify new feature ideas now that the current behavior is documented anyway.

Therefore I prefer to get this fix merged without waiting for the more involved additional development, and first separately [discuss ideas](https://github.com/RexOps/Rex/discussions/new?category=ideas) about proposals to change the current behavior.
